### ISSUE TITLE
Buffer size

### ISSUE THREAD

**Issue by LightTheSylveon** [ronf/asyncssh#329]
Hello, when I send a big buffer on the login screen (or anywhere else), the SSH server freezes and it is unusable. How could I fix this?

(If you didn't understand, If I send something really big to the server, the server freezes for like 10 seconds, depends on how big the string is, and the SSH server (the one I made with AsyncSSH) freezes completely.)

(code for SSHserver)

class MainSSHServer(asyncssh.SSHServer):

    def __init__(self):
        asyncssh.SSHServer.__init__(self)

    def connection_made(self, conn):
        pass

    def connection_lost(self, exc):
        if exc:
            print('SSH connection error: ' + str(exc), file=sys.stderr)
        else:
            print('SSH connection closed.')

    def password_auth_supported(self):
        return True

    def validate_password(self, username, password):
        return db.database.check_login(username, password)

async def start_server():
    await asyncssh.create_server(
        MainSSHServer, '', (port),
        server_host_keys=['ext/keys/id_rsa'],
        process_factory=handle_client
    )

**Comment by ronf** [comment]
Hello, I'm happy to try and help, but I'm not sure I understand what you're trying to do exactly.

You don't show what your handle_client() function is doing. Is it reading data from process.stdin? If not, there's a limit to how much most SSH clients will write without the other side reading, and that'll cause things to eventually block.

When you say "something really big", how many bytes are you talking about?

Also, I'm not sure what you mean about sending this data "on the login screen". Are you talking about the SSH auth banner feature, or something else?

Finally, you refer to sending to "the server", but you say you are implementing your own SSH server with AsyncSSH. So, which "server" are you sending data to, and what kind of channel is it being sent over?

**Comment by LightTheSylveon** [comment]
Sorry for being a bit unclear, I mean the SSH server with AsyncSSH, and I'm talking about SSH auth banner feature. And let's see how big we're talking about. I kind of just want to limit the buffer of what's allowed to be sent to the server. length of the bytes is roughly 16k bytes. But I still need the server to handle it.

async def handle_client(_conn):
    if _conn.get_extra_info('peername')[0] in settings['cnc_blacklists']:
        _conn.stdout.write("You are blacklisted from the CNC.\r\n")
    else:
        client = conn.Client(_conn.get_extra_info('peername')[0], _conn, _conn.get_extra_info('username'))
        client.address = _conn.get_extra_info('peername')
        conn.lookup.add_client(client)
    
        handler = handle.ClientHandler(client, _conn)
        await handler.start()

That's the handle_client() function

**Comment by ronf** [comment]
From the point of view of AsyncSSH and OpenSSH, 16 KB would not be considered a large amount of data. In fact, it would fit comfortably in a single SSH packet (which allows up to 32 KB by default). Flow control doesn't really become an issue until a much larger amount of data is sent (2 MB by default). So, that shouldn't be an issue between a typical SSH client and the AsyncSSH server. However, if AsyncSSH is forwarding requests to another server (particularly an embedded system of some kind), it's possible it could have much more restrictive limits.

Unfortunately, there's still not enough code here for me to tell what's going on. Since you are setting handle_client as your process factory, I would expect the argument to it to be of type SSHServerProcess, and that's consistent with your calls to functions like _conn.get_extra_info() and _conn.stdout.write(). I'm not sure what class "conn" or "handler" is, though, and I also don't see any mention of the auth banner feature or any writing of data to some other server.

If this is somehow related to the auth banner feature, I'm also not sure how this other server would be involved in that, as the auth banner would be sent from the AsyncSSH server to the SSH client.

**Comment by ronf** [comment]
I'm still not clear on exactly what you are doing here. Could you provide a more complete example script that shows the problem, or share a bit more of the code to show what you are doing with the data in spam.txt?

I took a look at the YouTube video, but it happened a bit too quickly for me to completely follow what you were doing. Were you sending the contents of spam.txt while sitting a the "login as:" prompt? If so, that prompt is generated by PuTTY, not the AsyncSSH server, and so I'm not sure if PuTTY even tries to send all that data to AsyncSSH. If it did, it would be part of a USERAUTH_REQUEST message, and those are not allowed to be that big. I'm not sure there's anything I could do in AsyncSSH to address that case.

Also, spam.txt is much larger than you mentioned -- it's over 3 MBytes. If you're sending this after the authentication, it's possible that this would cause flow control to kick in when forwarding the data. If the other end of the connection is not reading data at the time you send this, you could get into a deadlock where the client will freeze up waiting for the server to process the data it has sent before sending any more.

Is authentication completing successfully on your AsyncSSH server? If so, are you reading data from the client after it completes?

Have you tried enabling debug logging on AsyncSSH? The lines to do that would look something like:

```python
    import logging
    logging.basicConfig(level='DEBUG')
    asyncssh.set_debug_level(2)
```

If you could capture that output, that might provide a bit more information about what's going on.

**Comment by LightTheSylveon** [comment]
Sure! Also thanks for helping, even though I'm not being very clear sorry!, going to try that right now!

**Comment by LightTheSylveon** [comment]
Okay, so I set  the logging.

DEBUG:asyncio:Using selector: EpollSelector
INFO:asyncssh:Creating SSH listener on port 558
INFO:asyncssh:[conn=0] Accepted SSH connection on 51.195.51.28, port 558
INFO:asyncssh:[conn=0]   Client address: 10.66.66.2, port 50740
DEBUG:asyncssh:[conn=0] Requesting key exchange
DEBUG:asyncssh:[conn=0]   Key exchange algs: curve25519-sha256,curve25519-sha256@libssh.org,curve448-sha512,ecdh-sha2-nistp521,ecdh-sha2-nistp384,ecdh-sha2-nistp256,ecdh-sha2-1.3.132.0.10,diffie-hellman-group-exchange-sha256,diffie-hellman-group14-sha256,diffie-hellman-group15-sha512,diffie-hellman-group16-sha512,diffie-hellman-group17-sha512,diffie-hellman-group18-sha512,diffie-hellman-group14-sha1,rsa2048-sha256
DEBUG:asyncssh:[conn=0]   Host key algs: rsa-sha2-256,rsa-sha2-512,ssh-rsa
DEBUG:asyncssh:[conn=0]   Encryption algs: chacha20-poly1305@openssh.com,aes256-gcm@openssh.com,aes128-gcm@openssh.com,aes256-ctr,aes192-ctr,aes128-ctr
DEBUG:asyncssh:[conn=0]   MAC algs: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
DEBUG:asyncssh:[conn=0]   Compression algs: zlib@openssh.com,none
DEBUG:asyncssh:[conn=0] Received key exchange request
DEBUG:asyncssh:[conn=0]   Key exchange algs: curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group-exchange-sha1,diffie-hellman-group14-sha256,diffie-hellman-group14-sha1,rsa2048-sha256,rsa1024-sha1,diffie-hellman-group1-sha1
DEBUG:asyncssh:[conn=0]   Host key algs: ssh-rsa,ssh-ed25519,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,ssh-dss
DEBUG:asyncssh:[conn=0]   Client to server:
DEBUG:asyncssh:[conn=0]     Encryption algs: aes256-ctr,aes256-cbc,rijndael-cbc@lysator.liu.se,aes192-ctr,aes192-cbc,aes128-ctr,aes128-cbc,chacha20-poly1305@openssh.com,3des-ctr,3des-cbc,blowfish-ctr,blowfish-cbc,arcfour256,arcfour128
DEBUG:asyncssh:[conn=0]     MAC algs: hmac-sha2-256,hmac-sha1,hmac-sha1-96,hmac-md5,hmac-sha2-256-etm@openssh.com,hmac-sha1-etm@openssh.com,hmac-sha1-96-etm@openssh.com,hmac-md5-etm@openssh.com
DEBUG:asyncssh:[conn=0]     Compression algs: none,zlib,zlib@openssh.com
DEBUG:asyncssh:[conn=0]   Server to client:
DEBUG:asyncssh:[conn=0]     Encryption algs: aes256-ctr,aes256-cbc,rijndael-cbc@lysator.liu.se,aes192-ctr,aes192-cbc,aes128-ctr,aes128-cbc,chacha20-poly1305@openssh.com,3des-ctr,3des-cbc,blowfish-ctr,blowfish-cbc,arcfour256,arcfour128
DEBUG:asyncssh:[conn=0]     MAC algs: hmac-sha2-256,hmac-sha1,hmac-sha1-96,hmac-md5,hmac-sha2-256-etm@openssh.com,hmac-sha1-etm@openssh.com,hmac-sha1-96-etm@openssh.com,hmac-md5-etm@openssh.com
DEBUG:asyncssh:[conn=0]     Compression algs: none,zlib,zlib@openssh.com
DEBUG:asyncssh:[conn=0] Beginning key exchange
DEBUG:asyncssh:[conn=0]   Key exchange alg: curve25519-sha256@libssh.org
DEBUG:asyncssh:[conn=0]   Client to server:
DEBUG:asyncssh:[conn=0]     Encryption alg: aes256-ctr
DEBUG:asyncssh:[conn=0]     MAC alg: hmac-sha2-256
DEBUG:asyncssh:[conn=0]     Compression alg: none
DEBUG:asyncssh:[conn=0]   Server to client:
DEBUG:asyncssh:[conn=0]     Encryption alg: aes256-ctr
DEBUG:asyncssh:[conn=0]     MAC alg: hmac-sha2-256
DEBUG:asyncssh:[conn=0]     Compression alg: none
DEBUG:asyncssh:[conn=0] Completed key exchange
DEBUG:asyncssh:[conn=0] Accepting request for service ssh-userauth


(that's the output)

and when I insert the data from spam.txt, it doesn't send anything, until sometime later, and then it sends the contents of spam.txt, 

(video) https://youtu.be/j-VyD-FNQ7s

**Comment by ronf** [comment]
Ok - it definitely looks like it is using the large string you pass in as a username in SSH password authentication.

I see that you're doing some kind of database lookup to validate the password. If you temporarily take that out, do you still see the delay? Also, if you make "validate_password" an async function, do other users get locked out? Since it's not async right now, any delay in validate_password would end up locking up the entire AsyncSSH server.

It would also be good to see if you get the same delay with a different SSH client (like OpenSSH).

**Comment by LightTheSylveon** [comment]
Sure, let me take that out, one second.

**Comment by LightTheSylveon** [comment]
async def handle_client(_conn):
    pass
    #if _conn.get_extra_info('peername')[0] in settings['cnc_blacklists']:
    #    _conn.stdout.write("You are blacklisted from the CNC.\r\n")
    #else:
    #    client = conn.Client(_conn.get_extra_info('peername')[0], _conn, _conn.get_extra_info('username'))
    #    client.address = _conn.get_extra_info('peername')
    #    conn.lookup.add_client(client)
    #
    #    handler = handle.ClientHandler(client, _conn)
    #    await handler.start()

class MainSSHServer(asyncssh.SSHServer):

    def __init__(self):
        asyncssh.SSHServer.__init__(self)

    def connection_made(self, conn):
        pass

    def connection_lost(self, exc):
        if exc:
            print('SSH connection error: ' + str(exc), file=sys.stderr)
        else:
            print('SSH connection closed.')

    def password_auth_supported(self):
        return True

    def validate_password(self, username, password):
        return True
        #return db.database.check_login(username, password)

async def start_server():
    await asyncssh.create_server(
        MainSSHServer, '', 558,
        server_host_keys=['ext/keys/id_rsa'],
        process_factory=handle_client
    )


Using this code, still the exact same issue!

**Comment by ronf** [comment]
It will take some time to transfer that much data, so I would expect some delay no matter what here, but I was thinking the delay might be less without the database lookup.

If you change "def validate_password" to "async def validate_password", does that at least prevent the lockup for other users?

**Comment by LightTheSylveon** [comment]
Let me try that.

**Comment by LightTheSylveon** [comment]
async def validate_password(self, username, password):
        return True

I made this change, and it still messes it up for other users.

**Comment by ronf** [comment]
Don't change "password_auth_supported" to async -- only "validate_password".

**Comment by LightTheSylveon** [comment]
I tried that too, still it doesn't accept other clients

**Comment by ronf** [comment]
It would be interesting to see what the traceback looks like when you Ctrl-C while it is in the middle of hanging.

**Comment by LightTheSylveon** [comment]
second, I can show you that.

**Comment by LightTheSylveon** [comment]
^CTraceback (most recent call last):
  File "main.py", line 16, in <module>
    loop.run_forever()
  File "/usr/local/lib/python3.7/asyncio/base_events.py", line 534, in run_forever
    self._run_once()
  File "/usr/local/lib/python3.7/asyncio/base_events.py", line 1771, in _run_once
    handle._run()
  File "/usr/local/lib/python3.7/asyncio/events.py", line 88, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/local/lib/python3.7/asyncio/selector_events.py", line 767, in _read_ready
    self._read_ready_cb()
  File "/usr/local/lib/python3.7/asyncio/selector_events.py", line 816, in _read_ready__data_received
    self._protocol.data_received(data)
  File "/usr/local/lib/python3.7/site-packages/asyncssh/connection.py", line 910, in data_received
    while self._inpbuf and self._recv_handler():
  File "/usr/local/lib/python3.7/site-packages/asyncssh/connection.py", line 1143, in _recv_packet
    processed = handler.process_packet(pkttype, seq, packet)
  File "/usr/local/lib/python3.7/site-packages/asyncssh/packet.py", line 215, in process_packet
    self._packet_handlers[pkttype](self, pkttype, pktid, packet)
  File "/usr/local/lib/python3.7/site-packages/asyncssh/connection.py", line 1782, in _process_userauth_request
    username = saslprep(username.decode('utf-8'))
  File "/usr/local/lib/python3.7/site-packages/asyncssh/saslprep.py", line 114, in saslprep
    return _stringprep(s, True, _map_saslprep, 'NFKC', prohibited, True)
  File "/usr/local/lib/python3.7/site-packages/asyncssh/saslprep.py", line 70, in _stringprep
    if stringprep.in_table_a1(c):
  File "/usr/local/lib/python3.7/stringprep.py", line 13, in in_table_a1
    if unicodedata.category(code) != 'Cn': return False

**Comment by ronf** [comment]
Ah, interesting! It looks like it is getting stuck in the Unicode normalization code for usernames.

One possible workaround for this would be to limit the length of the username before attempting to normalize it.

**Comment by LightTheSylveon** [comment]
yes! that is what I was trying to do in the first place, how would I do it?

**Comment by LightTheSylveon** [comment]
I couldn't find any documentation on it on the docs :( (or im blind)

**Comment by ronf** [comment]
There's no way to do this today without a code change, but I can suggest something for you to try, and I can look into adding this check to the real release if it works.

Take a look in the _process_userauth_request function in connection.py. I'm thinking you'll need to add code after "method=packet.get_string()" that looks something like:

```python
    def _process_userauth_request(self, _pkttype, _pktid, packet):
        """Process a user authentication request"""

        username = packet.get_string()
        service = packet.get_string()
        method = packet.get_string()

        if len(username) >= 1024:                         # THIS CODE IS NEW
            raise IllegalUserName('Username too long')

        if service != _CONNECTION_SERVICE:
            raise ServiceNotAvailable('Unexpected service in auth request')

        try:
            username = saslprep(username.decode('utf-8'))
        except (UnicodeDecodeError, SASLPrepError) as exc:
            raise IllegalUserName(str(exc)) from None
```

**Comment by LightTheSylveon** [comment]
sure thing, could you maybe try to make it a feature in the library? I very much enjoying using this library, but it would make my life a lot easier that way :), like a buffer limit :), thank you very much though!

**Comment by ronf** [comment]
Yes - I'd definitely want to add this to the library if it solves your problem.

**Comment by LightTheSylveon** [comment]
Thank you so much!!! Very great support, thanks a lot man :), let me see if it solves my problem!

**Comment by LightTheSylveon** [comment]
It completely fixes my problem, thank you so much!!!

**Comment by ronf** [comment]
Terrific! Thanks for the confirmation, and for reporting the original problem.

I've got some other changes in progress right now on that file, but I'll add this in the "develop" branch as soon as those are committed, and that'll then make it into the next release.

**Comment by LightTheSylveon** [comment]
Sure thing!, thank you so much for the help, I've been struggling with this for a very long time :)

**Comment by LightTheSylveon** [comment]
Theres one more problem, it does this too for my custom shell I implemented using it, 

^CTraceback (most recent call last):
  File "main.py", line 16, in <module>
    loop.run_forever()
  File "/usr/local/lib/python3.7/asyncio/base_events.py", line 534, in run_forever
    self._run_once()
  File "/usr/local/lib/python3.7/asyncio/base_events.py", line 1771, in _run_once
    handle._run()
  File "/usr/local/lib/python3.7/asyncio/events.py", line 88, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/local/lib/python3.7/asyncio/selector_events.py", line 767, in _read_ready
    self._read_ready_cb()
  File "/usr/local/lib/python3.7/asyncio/selector_events.py", line 816, in _read_ready__data_received
    self._protocol.data_received(data)
  File "/usr/local/lib/python3.7/site-packages/asyncssh/connection.py", line 910, in data_received
    while self._inpbuf and self._recv_handler():
  File "/usr/local/lib/python3.7/site-packages/asyncssh/connection.py", line 1143, in _recv_packet
    processed = handler.process_packet(pkttype, seq, packet)
  File "/usr/local/lib/python3.7/site-packages/asyncssh/packet.py", line 215, in process_packet
    self._packet_handlers[pkttype](self, pkttype, pktid, packet)
  File "/usr/local/lib/python3.7/site-packages/asyncssh/channel.py", line 523, in _process_data
    self._accept_data(data)
  File "/usr/local/lib/python3.7/site-packages/asyncssh/channel.py", line 369, in _accept_data
    self._deliver_data(data, datatype)
  File "/usr/local/lib/python3.7/site-packages/asyncssh/channel.py", line 333, in _deliver_data
    self._session.data_received(data, datatype)
  File "/usr/local/lib/python3.7/site-packages/asyncssh/editor.py", line 710, in data_received
    self._editor.process_input(data, datatype)
  File "/usr/local/lib/python3.7/site-packages/asyncssh/editor.py", line 503, in process_input
    self._insert_printable(ch)
  File "/usr/local/lib/python3.7/site-packages/asyncssh/editor.py", line 261, in _insert_printable
    self._update_input(self._pos - data_len, self._pos)
  File "/usr/local/lib/python3.7/site-packages/asyncssh/editor.py", line 251, in _update_input
    self._reposition(new_pos)
  File "/usr/local/lib/python3.7/site-packages/asyncssh/editor.py", line 216, in _reposition
    self._move_cursor(new_pos)
  File "/usr/local/lib/python3.7/site-packages/asyncssh/editor.py", line 187, in _move_cursor
    new_column = self._column(pos)
  File "/usr/local/lib/python3.7/site-packages/asyncssh/editor.py", line 115, in _column
    self._start_column)
  File "/usr/local/lib/python3.7/site-packages/asyncssh/editor.py", line 104, in _determine_column
    if _is_wide(ch) and (column % self._width) == self._width - 1:
  File "/usr/local/lib/python3.7/site-packages/asyncssh/editor.py", line 38, in _is_wide
    return east_asian_width(ch) in 'WF'


// code for writing shell and reading
self.s.write(self.decoration['shellprompt'])
self.rdbuf = await self.conn.stdin.read(1024)
self.rdbuf = self.rdbuf.strip()

**Comment by ronf** [comment]
That appears to be in the line editor code. Is that a feature you need? If not, you could try turning that off by passing in "line_editor=False" when starting up the server.

**Comment by LightTheSylveon** [comment]
Let me see, one second

**Comment by LightTheSylveon** [comment]
That fixes it, thank you!, also how would I get keyboard input seperately, like arrow up, arrow down, etc.?

**Comment by ronf** [comment]
Without the line editor doing the handling of things like arrow keys for you, it would be a fair amount of work to get that input, as you'd still need to go into character-at-a-time mode on the client TTY so it doesn't do its own echo and input handling.

The issue could be addressed by setting a maximum allowed line length, but it's not quite as obvious how large to make that.

Regarding the pause_writing error, could you provide a full traceback?

**Comment by LightTheSylveon** [comment]
One second. I sure can

**Comment by LightTheSylveon** [comment]
I am somehow not getting that anymore, but my terminal bugs out. Let me show you:

![image](https://user-images.githubusercontent.com/55628225/100389978-3d998700-302f-11eb-90aa-6a0569d2f611.png)

That happens when I press enter, it should write it to a newline obviously.

**Comment by ronf** [comment]
You might need to change the places you output newlines from "\n" to "\r\n" if you disable the line editor. That's one of the things it does for you.

**Comment by LightTheSylveon** [comment]
yeah I thought of doing that, let me try that.

**Comment by ronf** [comment]
Providing a way to set a maximum line length on the editor to avoid it taking too long in _is_wide() is probably the better long-term answer, but that'll require some additional work in the editor and a new config parameter.

**Comment by LightTheSylveon** [comment]
yeah, this completely messes with my client, it doesn't even show what I type anymore with line_editor=False

**Comment by LightTheSylveon** [comment]
Could you help me with using _is_wide()?

**Comment by ronf** [comment]
Yeah - the line editor does echo as well. I guess you'd need to run the client with a TTY disabled if you wanted to turn the line editor off but still get local echo.

**Comment by LightTheSylveon** [comment]
oh okay, well another issue really quick, how would I use history? like with arrow key up.

**Comment by ronf** [comment]
You can't support history without enabling the line editor (or adding a bunch of code to implement that yourself). That's part of why it is there (along with the other things you mentioned here about echo and newlines).

There's no quick fix for someone sending too long a line to the line editor. You'd have to add code to limit the length. It wouldn't be a complete fix, but one thing that would get you close would be to change the code in process_input() to limit what gets added to the input buffer. Down near the bottom of that function, there's a block the looks like:

```python
                elif self._key_state == self._keymap and ch.isprintable():
                    self._insert_printable(ch)
```

This would need to add an additional check on the line length such as:

```python
                elif self._key_state == self._keymap and ch.isprintable() and len(self._line) < 1024:
                    self._insert_printable(ch)
```

A real version of this would make the 1024 a config value that's passed in, and probably allow it to still default to unlimited. Also, there are some other places that add data to the input line that would also need to be changed to have a complete solution, like the _insert_erased function.

**Comment by ronf** [comment]
Getting back to your arrow keys question, if the shell isn't echoing anything by default, it's probably already in character-at-a-time mode, so you'd have to look for multi-byte sequences for arrow and function keys.

For instance, up arrow is usually something like ESC ('\x1b`) followed by '[' and then 'A'. This means you need a state machine that doesn't do anything when it gets the first part of such a sequence, but remembers what was seen to know what to look for next, and then eventually calls the right handler once it gets a full sequence. That's what the "_keylist" and "_keymap" data structures are all about in the line editor code.

**Comment by LightTheSylveon** [comment]
ohh okay, thanks! I will try what you said above ^^

**Comment by LightTheSylveon** [comment]
Sorry for the late response!

**Comment by LightTheSylveon** [comment]
elif self._key_state == self._keymap and ch.isprintable():
                    self._insert_printable(ch)

This would need to add an additional check on the line length such as:

                elif self._key_state == self._keymap and ch.isprintable() and len(self._line) < 1024:
                    self._insert_printable(ch)


Hey, I made this change, but it still messes up the asyncSSH server. I am getting an exception

socket.send() raised exception.
socket.send() raised exception.
socket.send() raised exception.
socket.send() raised exception.


Please help?

**Comment by ronf** [comment]
Did you turn back on the line editor after making this change? What sort of failure are you seeing after that?

Limiting the length of the input line should reduce the cost, but it is still going to take a long time to process several megabytes of pasted data, especially with it trying to ring the bell on your terminal in response to each character. There are some things that I might not be able to fully protect against.

As an example, when I tried a similar experiment of pasting into a Terminal.app window running "zsh", that particular terminal window "froze" with no echo of the pasted data and without accepting any new input for several minutes, though in the end it did actually eventually come back to life and show the input I pasted.

**Comment by LightTheSylveon** [comment]
Well I tried reducing the length, but it has no effect :/, anything else I can try?

**Comment by LightTheSylveon** [comment]
Just want to say I fixed it myself ! it was a very very easy fix! 
    def process_input(self, data, datatype):
        """Process input from channel"""
        
        if len(data) > 1024 or len(self._line) > 1024:
          return

        if self._line_mode:
            for ch in data:
                if ch in self._key_state:
                    self._key_state = self._key_state[ch]
                    if callable(self._key_state):
                        try:
                            self._key_state(self)
                        finally:
                            self._key_state = self._keymap
                elif self._key_state == self._keymap and ch.isprintable():
                    self._insert_printable(ch)
                else:
                    self._key_state = self._keymap
                    self._ring_bell()
        else:
            self._session.data_received(data, datatype)

**Comment by ronf** [comment]
Thanks for passing that on.

While this will prevent something like a large paste from getting through, it does so by simply throwing away large blocks of input whenever they are received. This could cause problems if you had an automated client connecting to your AsyncSSH server which sent a large block of data, even if no individual line exceeds 1024 characters. If there's a newline in that block of data, you'll never look for it, missing your chance to end the current line. This may be ok in your use-case if you are only expecting interactive clients, but it's not something I could include in the library itself.

I'll continue thinking about this and let you know if I come up with anything, but some other changes I'm working on are taking precedence over this right now.

**Comment by LightTheSylveon** [comment]
Sure thing! thank you!

**Comment by ronf** [comment]
The username length limit change we discussed is now available as commit 8c2546f in the "develop" branch and will be part of the next release. I'm still thinking about potential line editor changes, but wanted to get this fix in place in the meantime.

**Comment by LightTheSylveon** [comment]
Perfect!!!!!!

**Comment by ronf** [comment]
Ok - I've started to look at this more closely. There are definitely multiple changes needed to properly handle a very large paste of this sort, and I'm not sure in the end if I'll be able to fully address all possible edge cases. However, there are definitely improvements that can be made here that should help.

First, I noticed that even after the client doing the large paste drops the connection, the server continues to process the leftover input and attempt to echo it back, leaving all other clients hung. This is relatively simple to fix, and should be useful in other circumstances when the client drops the connection unexpectedly. Among other things, this should prevent or greatly reduce the "socket.send() raised exception." messages output by asyncio when trying to write to a broken connection. Here's the fix:

```patch
diff --git a/asyncssh/connection.py b/asyncssh/connection.py
index d30a928..449ee63 100644
--- a/asyncssh/connection.py
+++ b/asyncssh/connection.py
@@ -997,6 +997,9 @@ class SSHConnection(SSHPacketHandler, asyncio.Protocol):
         """Send data to the SSH connection"""
 
         if self._transport:
+            if self._transport.is_closing():
+                raise BrokenPipeError
+
             self._transport.write(data)
 
     def _send_version(self):
```

With this change, as soon as the client doing the large paste drops the connection the server should immediately recover, allowing other clients to make progress again.

Another issue is that right now the process_input() function is a synchronous callback, not a coroutine, and it has to process all the data provided to it before it returns to the event loop. The same goes for process_output(). Because these are not async, it's not just a matter of adding something like an "await asyncio.sleep(0)" in them to periodically yield to other coroutines. I'm not sure how straightforward that will be to change, but that's what I'll be looking at next.

**Comment by ronf** [comment]
Quick update: I've made some progress on this, but it's still not working as well as I'd like. Once I have something more complete, I'll make it available for you to try.

**Comment by LightTheSylveon** [comment]
Hey!, sorry I haven't responded the last few days! Got tested for covid-19 and it's all been a bit of a mess, apologies!,  thanks for the amount of progress you made, I'll be sure to check it out when you finish it, thanks a lot!

**Comment by ronf** [comment]
Sounds scary - hope everything works out for you!

I'm still working on this, but I probably won't have anything ready until the weekend at the earliest. I'm going to focus on just optimizing the code for now, and not on trying to split up the work into multiple callbacks. Between that and allowing a max input length to be set, I think it should solve the problem you're having, and I can revisit whether or not to break it into multiple callbacks later.

**Comment by LightTheSylveon** [comment]
Sounds good!

**Comment by ronf** [comment]
I finally had a chance to get back to this. The optimizations were a little more involved than I would have liked, and they're still not enough to allow input lines which are megabytes in size, but I did managed to speed things up somewhat. Before the change, even a 16 KB input line took something like 1-2 minutes to process, where the new code (with no length limit set) only takes about 0.1-0.2 seconds (about 600x faster), and even going to 128 KB with no length limit only takes about 1 second and pretty much all of that time is in making the SSH connection, not in the input processing.

With a length limit set, things get even faster. Even with the optimized code, running the original 3 MB input line takes like 6-7 minutes with no length limit, but with the default 1024 byte length limit, it completes in 2-3 seconds.

I have one more idea that I'd like to try that I think could speed things up even more, but I thought this set of changes may be good enough for you to use in the meantime. So, I've gone ahead and checked this into the "develop" branch as commits 810bc43 and d0f0725.

**Comment by ronf** [comment]
Ok - with the additional optimization I just added, even the 3 MB input line with no length limit takes under 2 seconds to enter and echo back (an additional ~200x speed-up on top of the previous improvements), and with the default 1024-byte length limit it takes about 0.1 seconds to process that data. This change only helps when the majority of the input is printable ASCII characters, so it won't currently help as much with something like all Kanji input, but I think it's still a good improvement.

With all of these changes, there are still some edge cases where you might see some delays with pathological input, but it should be much improved, particularly with the new default max line length of 1024 characters.

This change is now available as commit b83cb83 in the "develop" branch.

**Comment by LightTheSylveon** [comment]
Thanks man, ill start using that, cheers!, also I experienced some issues where I had to use stderr instead of stdout, and it magically works, ALSO! if I use the code I showed above, the server will not disconnect people that unplug their internet.

**Comment by ronf** [comment]
The 1024 limit you put into your change is now baked into the changes I committed here, so you shouldn't need to do anything to get that behavior now. If you want to make it larger (or unlimited) you can do so using the new "max_line_length" argument when creating a server, but AsyncSSH will limit the input to 1024 characters by default.

I'm not quite sure what you mean here about stdout vs. stderr. Keep in mind that the SSH protocol doesn't provide "stderr" when a client opens a session and requests a pseudo-TTY, and the line editor only gets enabled when a pseudo-TTY is requested. So, if you try to write something to stderr, it should end up showing up on the client no differently than anything you write to stdout. With a pseudo-TTY allocated, they aren't separate data streams any more.

**Comment by ronf** [comment]
I spent a bit more time on this today, and extended the printable ASCII optimization to cover a bunch of the Unicode space as well. I've limited it to only characters up through U+FFFF for now to keep the cost of building the matching regex from getting too large when a new line editor is instantiated, but that should cover the large majority of actual input this will get, and the remaining character still work (just using the slow path). Also, it's a one-line change to extend it, if that's ever needed.
